File: 백준/레벨Gold III/1005. ACM Craft/ACM Craft.py
from sys import stdin

input = stdin.readline

N, K, W = 0, 0, 0
build_time = []
graph = []
in_degree = []

def topological_sort():
    stk = []
    dp = [0] * (N + 1)

    # 큐, DP 초기화
    for i in range(1, N + 1):
        if in_degree[i] == 0:
            stk.append(i)
            dp[i] = build_time[i]

    # 위상 정렬
    while stk:
        u = stk.pop()

        for v in graph[u]:
            in_degree[v] -= 1
            dp[v] = max(dp[v], dp[u] + build_time[v])
            
            if in_degree[v] == 0:
                stk.append(v)

    return dp

def test():    
    global N, K, build_time, graph, in_degree, W

    # 빌드타임 초기화
    N, K = map(int, input().split())
    build_time = [0] * (N + 1)

    for idx, time in enumerate(map(int, input().split()), start=1):
        build_time[idx] = time

    # 인접리스트, 진입차수 초기화
    graph = [[] for _ in range(N + 1)]
    in_degree = [0] * (N + 1)
    
    for _ in range(K):
        u, v = map(int, input().split())
        graph[u].append(v)
        in_degree[v] += 1

    # 승리조건
    W = int(input().rstrip())

    # 위상정렬
    dp = topological_sort()

    return dp[W]

def solution():    
    T = int(input().rstrip())
    
    for _ in range(T):
        print(test())

solution()

File: 백준/레벨Gold III/1005. ACM Craft/ACM Craft.ts
// @ts-ignore
import fs = require('fs');

const lines = fs.readFileSync('/dev/stdin').toString().split('\n');
let lineIndex = 0;
const input = (): string => lines[lineIndex++];

let N = 0;
let K = 0;
let W = 0;
let buildTime: number[] = [];
let graph: number[][] = [];
let inDegree: number[] = [];

const topological_sort = (): number[] => {
  const stk: number[] = [];
  const dp = Array(N + 1).fill(0);

  // 큐, DP 초기화
  for (let i = 1; i <= N; i++) {
    if (inDegree[i] === 0) {
      stk.push(i);
      dp[i] = buildTime[i];
    }
  }

  // 위상 정렬
  while (stk.length > 0) {
    const u = stk.pop()!;

    for (const v of graph[u]) {
      inDegree[v]--;
      dp[v] = Math.max(dp[v], dp[u] + buildTime[v]);

      if (inDegree[v] === 0) {
        stk.push(v);
      }
    }
  }

  return dp;
};

const test = (): number => {
  // 빌드타임 초기화
  [N, K] = input().split(' ').map(Number);
  buildTime = [0, ...input().split(' ').map(Number)];

  // 인접리스트, 진입차수 초기화
  graph = Array.from({ length: N + 1 }, () => []);
  inDegree = Array(N + 1).fill(0);

  for (let i = 0; i < K; i++) {
    const [u, v] = input().split(' ').map(Number);
    graph[u].push(v);
    inDegree[v]++;
  }

  // 승리 조건
  W = Number(input().trim());

  // 위상 정렬
  const dp = topological_sort();

  return dp[W];
};

const solution = (): void => {
  const T = Number(input().trim());
  const result: number[] = [];

  for (let t = 0; t < T; t++) {
    result.push(test());
  }

  console.log(result.join('\n'));
};

solution();